doc/compiler/memory/MemoryManager.md

```diff
@@ -0,0 +1,214 @@
+# OMR Compiler Memory Manager
+
+The OMR Compiler doesn't necessarily operate in a vacuum, but rather 
+as one component of a Managed runtime. Additionally,  communications 
+between components could be extern "C". Therefore, one can't simply 
+use something like `::operator new` since it can throw. Additionally, 
+mutli-threaded performance of malloc historically has been hit or miss, 
+and any generic common allocator would have poor interaction between 
+components, with little to no insight into usage.
```

Comment:
sp. multi

I don't think throwing shade at `malloc` here is a good call - If `malloc`'s threaded performance were really at issue, there would be other allocators we could be deploying instead. The fact we don't is because we believe that we can derive substantial gains in both performance and functionality by having a compiler-domain-specific allocator. 

(I object mildly to the description of the port allocator as 'domain-specialized', because in the context of the compiler, it is precisely _not_ domain specific)

Reply:
I reworded the line regarding the port allocator to
>Using an allocator specific to a  Managed Runtime, such as `omrmem_allocate_memory` in the Port Library...

Regarding the comment about `malloc`, I think the point there is that from (our) historical empirical observation, `malloc` has been wasteful for memory when there are multiple simultaneous compilations occurring (since it uses per thread arenas + a global arena); it also results in more fragmentation since memory allocated in a compilation gets freed at the end of the compile but `malloc` doesn't release that memory back to the OS. In terms of other allocators we could be deploying, that's addressed by the line
>any generic common allocator would have poor interaction between components, with little to no insight into usage.

So it's not so much that we're putting down malloc so much as we're stating that in certain compiler configurations, `malloc` ends up hurting more than helping.

I can update the doc to be more explicit about what "hit or miss" means, or I can still take it out if you think it doesn't belong there.

Reply:
:D I think putting in the above is a pretty good call -- but at the end of the day, we have a story where we believe we can do pretty well for our allocation patterns by taking matters into our own hands.

Reply:
Re: shade on malloc. I thing the odds are high that a large part of why we went down this road is because malloc's _historical_ multi-threaded performance was awful.  A lot of this work started when dlmalloc was still the leading implementation.  I have huge respect for dlmalloc, but it is _not_ thread friendly.  Whether or not the historical context belongs in document describing current practice is :shrug: from me.

malloc's multi-threaded performance notwithstanding.  The reason this is all justified today, is that the compiler produces a prolific amount of garbage during a compilation, and if it were all allocated individually from a generic allocator, we'd be at serious risk of long-term memory fragmentation.  That's on top of any heavy-weight RAS requirements for a common runtime allocator.